File: app/src/main/java/com/banuba/sdk/example/quickstart_c_api/OffscreenEffectPlayer.java
package com.banuba.sdk.example.quickstart_c_api;

import java.nio.ByteBuffer;

class OffscreenEffectPlayer {
    private long mOep = 0;
    private DataReadyCallback mDataReadyCallback = null;
    private static boolean isInit = false;

    public static void init(String pathToResources, String clientToken) {
        assert !isInit;
        if (!isInit) {
            externalInit(pathToResources, clientToken);
            isInit = true;
        }
    }

    public static void deinit() {
        assert isInit;
        if (isInit) {
            isInit = false;
            externalDeinit();
        }
    }

    public OffscreenEffectPlayer(int width, int hegiht) {
        assert isInit;
        mOep = externalCreate(width, hegiht);
        surfaceChanged(width, hegiht);
    }

    @Override
    protected void finalize() {
        destroy();
    }

    public void destroy() {
        if (mOep != 0) {
            externalDestroy(mOep);
            mOep = 0;
        }
    }

    public void processImageAsync(ByteBuffer image, int width, int height, int inputRotation, boolean requireMirroring, int outputOrientation) {
        externalProcessImageAsync(mOep, image, width, height, inputRotation, requireMirroring, outputOrientation);
    }

    public void surfaceChanged(int width, int height) {
        externalSurfaceChanged(mOep, width, height);
    }

    public void loadEffect(String effectPath) {
        externalLoadEffect(mOep, effectPath);
    }

    public void unloadEffect() {
        externalUnloadEffect(mOep);
    }

    public void pause() {
        externalPause(mOep);
    }

    public void resume() {
        externalResume(mOep);
    }

    public void stop() {
        externalStop(mOep);
    }

    public void evalJs(String script) {
        externalEvalJs(mOep, script);
    }

    public interface DataReadyCallback {
        void onDataReady(byte[] image, int width, int height);
    }

    public void setDataReadyCallback(DataReadyCallback callback) {
        mDataReadyCallback = callback;
    }

    private void onDataReady(byte[] image, int width, int height) {
        if (mDataReadyCallback != null) {
            mDataReadyCallback.onDataReady(image, width, height);
        }
    }

    /* The functions below are implemented in c++ */
    private static native void externalInit(String pathToResources, String clientToken);
    private static native void externalDeinit();
    private native long externalCreate(int width, int height);
    private native void externalDestroy(long oep);
    private native void externalProcessImageAsync(long oep, ByteBuffer image, int width, int height, int inputRotation, boolean requireMirroring, int outputOrientation);
    private native void externalSurfaceChanged(long oep, int width, int height);
    private native void externalLoadEffect(long oep, String effectPath);
    private native void externalUnloadEffect(long oep);
    private native void externalPause(long oep);
    private native void externalResume(long oep);
    private native void externalStop(long oep);
    private native void externalEvalJs(long oep, String script);

    static {
        System.loadLibrary("native-lib");
    }
}

File: app/src/main/java/com/banuba/sdk/example/quickstart_c_api/OffscreenEffectPlayer.kt
package com.banuba.sdk.example.quickstart_c_api

import java.nio.ByteBuffer

class OffscreenEffectPlayer {
    private var callbackReady: ((ByteArray, Int, Int) -> Unit)? = null
    private var oep: Long = 0

    fun create(pathToResources: String, clientToken: String) {
        this.oep = externalCreate(pathToResources, clientToken)
    }

    fun destroy() {
        externalDestroy(this.oep)
        this.oep = 0
    }

    fun processImageAsync(image: ByteBuffer, width: Int, height: Int) {
        externalProcessImageAsync(this.oep, image, width, height)
    }

    fun surfaceChanged(width: Int, height: Int) {
        externalSurfaceChanged(this.oep, width, height)
    }

    fun loadEffect(name: String) {
        externalLoadEffect(this.oep, name)
    }

    fun unloadEffect() {
        externalUnloadEffect(this.oep)
    }

    fun pause() {
        externalPause(this.oep);
    }

    fun resume() {
        externalResume(this.oep)
    }

    fun callJsMethod(method: String, param: String) {
        externalCallJsMethod(this.oep, method, param)
    }

    fun setDataReadyCallback(callback: (image: ByteArray, width: Int, height: Int) -> Unit) {
        callbackReady = callback
    }

    fun dataReady(image: ByteArray, width: Int, height: Int) {
        callbackReady?.let { it(image, width, height) }
    }


    /* The functions below are implemented in c++ */
    private external fun externalCreate(pathToResources: String, clientToken: String): Long
    private external fun externalDestroy(oep: Long)
    private external fun externalProcessImageAsync(oep: Long, image: ByteBuffer, width: Int, height: Int)
    private external fun externalSurfaceChanged(oep: Long, width: Int, height: Int)
    private external fun externalLoadEffect(oep: Long, name: String)
    private external fun externalUnloadEffect(oep: Long)
    private external fun externalPause(oep: Long)
    private external fun externalResume(oep: Long)
    private external fun externalCallJsMethod(oep: Long, method: String, param: String)

    companion object {
        /* Used to load the 'native-lib' library on application startup. */
        init {
            System.loadLibrary("native-lib")
        }
    }
}
